R/separate-wider.R

```diff
@@ -0,0 +1,598 @@
+      i = 'Use `too_few = "debug"` to diagnose the problem.',
+      i = 'Use `too_few = "start"` to silence this message.'
+    ),
+    advice_long = c(
+      i = 'Use `too_many = "debug"` to diagnose the problem.',
+      i = 'Use `too_many = "drop"` to silence this message.'
+    ),
+    call = error_call
+  )
+
+  skip <- names(widths) == ""
+  from <- cbind(start = breaks[!skip], length = widths[!skip])
+  names <- names(widths)[!skip]
+
+  pieces <- stringr::str_sub_all(x, from)
+  pieces <- lapply(pieces, function(x) x[x != ""])
+
+  out <- df_align(
+    x = pieces,
+    names = names,
+    align_direction = if (too_few == "end") "end" else "start"
+  )
+
+  if (!col_remove || too_few == "debug" || too_many == "debug") {
+    out[[col]] <- x
+  }
+
+  if (too_few == "debug" || too_many == "debug") {
+    separate_warn_debug(col, names_sep, c("ok", "width", "remainder"))
+    problem <- !is.na(x) & (
+        (too_few == "debug" & width < expected_width) |
+        (too_many == "debug" & width > expected_width)
+      )
+
+    out[[debug_name(col, names_sep, "width")]] <- width
+    out[[debug_name(col, names_sep, "remainder")]] <- stringr::str_sub(x, expected_width + 1, width)
+    out[[debug_name(col, names_sep, "ok")]] <- !problem
+  }
+
+
+  out
+}
+
+#' @rdname separate_wider_delim
+#' @param patterns A named character vector where the names become column names
+#'   and the values are regular expressions that match the contents of the
+#'   vector. Unnamed components will match, but not be included in the output.
+#' @export
+separate_wider_regex <- function(
+    data,
+    cols,
+    patterns,
+    ...,
+    names_sep = NULL,
+    names_repair = "check_unique",
+    too_few = c("error", "debug", "start"),
+    col_remove = TRUE
+) {
+  check_installed("stringr")
+  check_required(cols)
+  check_character(patterns)
+  if (all(!have_name(patterns))) {
+    cli::cli_abort("{.arg patterns} must be a named character vector.")
+  }
+  check_dots_empty()
+  check_string(names_sep, allow_null = TRUE)
+  too_few <- arg_match(too_few)
+  check_bool(col_remove)
+
+  error_call <- current_env()
+
+  map_unpack(
+    data, {{ cols }},
+    function(x, col) str_separate_wider_regex(x, col,
+      patterns = patterns,
+      names_sep = names_sep,
+      too_few = too_few,
+      col_remove = col_remove,
+      error_call = error_call
+    ),
+    names_sep = names_sep,
+    names_repair = names_repair
+  )
+}
+
+str_separate_wider_regex <- function(x,
+                                     col,
+                                     patterns,
+                                     names_sep = NULL,
+                                     too_few = "error",
+                                     col_remove = TRUE,
+                                     error_call = caller_env()) {
+  has_name <- names2(patterns) != ""
+  groups <- paste0("(", ifelse(has_name, "", "?:"), patterns, ")")
+  full_match <- paste0("^", paste(groups, collapse = ""), "$")
```

Comment:
On the off case that you want to support `patterns = character()`, going to get burned by `paste0()` recycling

Reply:
🤮